server/src/main/java/org/opensearch/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,357 @@
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+
+        // TODO add handling for master switch, dangling indices, master reload
+
+        if(event.routingTableChanged()) {
+            updateIndexShardStatus(event.state());
+        }
+    }
+
+    private void updateIndexShardStatus(ClusterState clusterState) {
+        List<ShardRouting> routingTable;
+        Set<Index> relocationCompletedIndices = new HashSet<>();
+        Set<Index> failedIndices = new HashSet<>();
+        for(Index index: indexShardTieringStatus.keySet()) {
+            try {
+                // Ensure index is not deleted
+                routingTable = clusterState.routingTable().allShards(index.getName());
+            } catch (IndexNotFoundException ex) {
+                // Index already deleted nothing to do
+                logger.warn("Index [{}] deleted before hot to warm relocation finished", index.getName());
+                return;
+            }
+
+            Map<ShardId, ShardTieringStatus> shardTieringStatusMap = indexShardTieringStatus.getOrDefault(index, new HashMap<>());
+            List<ShardId> processingShards = shardTieringStatusMap.keySet().stream()
+                .filter(shardId -> shardTieringStatusMap.get(shardId).state() == State.INIT || shardTieringStatusMap.get(shardId).state() == State.PROCESSING)
+                .collect(Collectors.toList());
+            if (processingShards.isEmpty()) {
+                // No shards are in processing state, nothing to do
+                // This means that tiering for the index is completed - either failed or successful
+                continue;
+            }
+            boolean relocationCompleted = true;
+            for (ShardRouting shard : routingTable) {
+                if (shardTieringStatusMap.get(shard.shardId()) != null &&
+                    (State.SUCCESSFUL.equals(shardTieringStatusMap.get(shard.shardId()).state()) || State.FAILED.equals(shardTieringStatusMap.get(shard.shardId()).state()))) {
+                    continue;
+                }
+                TieringService.State  tieringState;
+                String reason = null;
+                boolean isShardFoundOnSearchNode = clusterState.getNodes().get(shard.currentNodeId()).isSearchNode();
+                boolean isShardRelocatingToSearchNode = clusterState.getNodes().get(shard.relocatingNodeId()).isSearchNode();
+
+                if (shard.active() && isShardFoundOnSearchNode) {
+                    tieringState = State.SUCCESSFUL;
+                } else if (isShardFoundOnSearchNode || isShardRelocatingToSearchNode) {
+                    tieringState = State.PROCESSING;
+                    relocationCompleted = false;
+                } else {
+                    tieringState = State.FAILED;
+                    relocationCompleted = false;
```

Comment:
This logic doesn't seem correct. The `routingTable` has all the `ShardRouting` which includes primary and replicas of all the shards. Whereas the tiering state of the shard is changed based on any the last `ShardRouting` state for that shard. So for a shard say if primary is moved but replica is not and in the list replica ShardRouting comes first followed by primary, we will mark it as complete for that shard.

Reply:
no thats not the case here. `relocationCompleted` is a boolean variable set to true by default, now the logic iterates through all the shards of the index through the `routingTable` and marks this variable as false if it finds any failure/processing state. `relocationCompleted` doesnt get set to true again, its set to true per index only one time.
Hope that clarifies.